### DISCUSSION TITLE
How to pass empty string `''` to `vim.fn.jobstart` in `cmd` parameter ?

### DISCUSSION THREAD

**Discussion by linrongbin16** [neovim/neovim#24598] [Q&A]
Hi, I want to pass this command:

```bash
rg --column -n --no-heading --color=always -S -- ''
```

to `vim.fn.jobstart(cmd, {...})` API, but it fails.

It seems it's because the last empty string `''` cannot been evaluate correctly, is there anything I could do for this???

device info:
OS: macOS M1 MAX, 13.5
Neovim: v0.9.1 Release, LuaJIT 2.1.0-beta3

**Comment by zeertzjq** [comment]
Repeat each single quote twice.

**Reply by linrongbin16** [reply]
hi @zeertzjq, is there any document about the reason? (I just want to learn it, because I want to write a script to dynamically pass user input to terminal inside nvim)

In my knowledge, I think I should pass `'\'\''` or `"''"` to cmd, because the single quotes inside a string should prepend with a backslash? is that correct or wrong?

**Reply by zeertzjq** [reply]
`:h literal-string`

**Reply by linrongbin16** [reply]
Hi @zeertzjq , I tried to use list instead of string:

This is the lua code:

```lua

local jobid = vim.fn.jobstart(
    { "rg", "--column", "-n", "--no-heading", "--color=always", "-S", "--", "''" },
    {
        on_stdout = function(chanid, data, name)
            print("stdout, data:" .. vim.inspect(data))
        end,
        on_stderr = function(chanid, data, name)
            print("stderr, data:" .. vim.inspect(data))
        end,
        on_exit = function(id, exitcode, event)
            print("exit, exitcode:" .. vim.inspect(exitcode))
        end,
    }
)
vim.fn.jobwait({ jobid })

```

<img width="787" alt="image" src="https://github.com/neovim/neovim/assets/6496887/b7a89905-248a-48e0-b08a-966441193257">

But it still not working, I add some `print` in the `on_stdout`, `on_stderr` and `on_exit`, but none of them really prints.

**Reply by linrongbin16** [reply]
While I found lua's `io.execute(cmd)` could work correctly (when the cmd is a string), there's no need to deal with escaping characters. 

But lua io cannot handle async IO, I want faster respond `on_stdout` to processing the command output when running command on big code base.

**Reply by linrongbin16** [reply]
I finally use the lua `io.popen` to solve this:

share the lua code:

```lua
local f = io.popen("rg --column -n --no-heading --color=always -S -- ''")
for line in f:lines("*line") do
   -- processing each line from stdout
end
f:close()
```
